specifyweb/frontend/js_src/lib/components/Statistics/hooks.tsx

```diff
@@ -334,53 +286,136 @@ export function useStatValueLoad<
   }, [promiseGenerator, value, onLoad]);
 }
 
+export function applyStatBackendResponse(
+  backEndResponse: BackendStatsResult,
+  items: RA<CustomStat | DefaultStat>,
+  responseKey: string,
+  formatter: (rawResult: any) => string | undefined,
+  statsSpec: StatsSpec
+): RA<CustomStat | DefaultStat> {
+  const phantomItem = items.find(
+    (item) =>
+      item.type === 'DefaultStat' &&
+      item.itemName === 'phantomItem' &&
+      item.pathToValue === undefined
+  );
+
+  const phantomItemUrlPrefix =
+    phantomItem === undefined || phantomItem.type !== 'DefaultStat'
+      ? undefined
+      : statsSpec[phantomItem.pageName].urlPrefix;
+
+  const phantomItemResponseKey =
+    phantomItemUrlPrefix === undefined
+      ? undefined
+      : generateStatUrl(
+          phantomItemUrlPrefix,
+          (phantomItem as DefaultStat).categoryName,
+          (phantomItem as DefaultStat).itemName
+        );
+  const isMyResponse = phantomItemResponseKey === responseKey;
+  return phantomItem !== undefined &&
+    isMyResponse &&
+    phantomItem.type === 'DefaultStat'
+    ? Object.entries(backEndResponse[responseKey]).map(
+        ([itemName, rawValue]) => ({
+          type: 'DefaultStat',
+          pageName: phantomItem.pageName,
+          itemName: 'phantomItem',
+          categoryName: phantomItem.categoryName,
+          label: itemName,
+          itemValue: formatter(rawValue),
+          itemType: 'BackEndStat',
+          pathToValue: itemName,
+        })
+      )
+    : items;
+}
+
+export function useDefaultDynamicCategorySetter(
+  defaultBackEndResponse: BackendStatsResult | undefined,
+  setDefaultLayout: (
+    previousGenerator: (
+      oldLayout: RA<StatLayout> | undefined
+    ) => RA<StatLayout> | undefined
+  ) => void
+) {
+  React.useEffect(() => {
+    dynamicStatsSpec.forEach(({ responseKey, formatter }) => {
+      if (
```

Comment:
(optional) To reduce code nesting, instead of `if(condition) { etc }` could do `if(!condition) return; etc`

Also, try to omit `{` and `}` when `if` has just one statement - curly braces are not vert readable and are easy to mistype - the fewer of them the better